ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,118 @@
+(*
+ * Copyright (C) 2006-2009 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Helpers
+
+module D = Debug.Make (struct let name = "tar_ext" end)
+
+open D
+
+let file_perm = 0o644
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let ( + ) = Int64.add
+
+exception Illegal_file_path of string
+
+exception Unsupported_file_type of string
+
+exception File_exceeds_max_size_limit of string
+
+exception File_incomplete
+
+exception File_parsed_failed
+
+let unpack dir acc_size max_size ifd head =
+  let assert_file_path_is_legal file_path =
+    if
+      List.exists
+        (fun s -> String.equal "." s || String.equal ".." s)
+        (String.split_on_char '/' file_path)
+    then
+      raise (Illegal_file_path file_path)
+  in
+  let assert_file_not_exceed_max_size acc_size =
```

Comment:
so we keep track of the max_size supplied by the caller. I presume this is not to protect against malicious user trying to unpacking a huge file then? What would be the purpose of keeping track of the max size?

Reply:
I see `max_size` is defined in xapi_globs:
`ocaml/xapi/xapi_globs.ml:948:let bundle_file_max_size = ref (Int64.of_int (500 * 1024 * 1024))`

Reply:
I'm not sure if there is a different "max size" case now and in future. I updated to check the limit directly in this module. If there is a different "max size" case, then will update it.